### ISSUE TITLE
make "required" carriers match their calling patterns

### ISSUE THREAD

**Issue by bensigelman** [opentracing/opentracing.github.io#73]
Right now we have two "required" carriers: SplitBinaryCarrier and SplitTextCarrier. The current spec snapshot describes them [here](https://github.com/opentracing/opentracing.github.io/blob/87a97c17fa159ff5401d89e6b5c8c496374274d5/spec.md#required-injectjoin-formats). I'll copy in the relevant passage:

> #### Required Inject/Join formats
> 
> At a minimum, all platforms require OpenTracing implementations to support two formats: the "split text" format and the "split binary" format. The "split" refers to two common components of propagated spans:
> 
> 1. The core identifying information for the `Span`, referred to as the "tracer state" (for example, in Dapper this would include a `trace_id`, a `span_id`, and a bitmask representing the sampling status for the given trace)
> 1. Any Baggage (per `Span`'s ability to set Baggage items that propagate across process boundaries)
> 
> The "text" and "binary" designations refer to two flavors of encoding:
> 
> - The *text* format is a platform-idiomatic map from (unicode) `string` to `string`
> - The *binary* format is an opaque byte array (and presumably more compact and efficient)
> 
> Note that there is no expectation that different tracing systems Inject and Join `Spans` in compatible ways. Though OpenTracing is agnostic about the tracing implementation, for successful inter-process handoff it's essential that the processes on both sides of a propagation use the same tracing implementation.

The carriers are "split" so that

1. OpenTracing users can opt-out of propagating baggage, and also so that
1. OpenTracing users can put the (potentially fixed-length) tracer state in a different place than the baggage in their respective wire formats.

The former can be accomplished via a flag or other control mechanism on the carrier or on the inject call. The latter doesn't really make sense in the text format, and really derives from a very particular requirement that Uber has, and it may make more sense for TChannel to define its own Carrier which it will prefer (and which Uber's OT implementation can support).

As such, here's a new proposal for the bulit-in carriers... the intention here is to make sure that the user can (a) always opt-out of Baggage, and (b) fit in with either HTTP headers or binary protocols.

I'll sketch these out in Go, but any language would do:
```
type MixedMapCarrier struct {
    // Inject() should include / Join() should extract Baggage items in
    // MixedMap iff BaggageEnabled==true.
    BaggageEnabled bool  // tempted to call this "CarryBaggage" :)

    // A map in which to store Tracer State and – depending on
    // BaggageEnabled – Baggage items.
    //
    // IMPORTANT: This map is *not* only used by the carrier... it may have
    // entries devoted to entirely unrelated things (e.g., arbitrary HTTP
    // Headers) and, as such, any carrier items must be distinguishable (e.g.,
    // via a shared prefix string).
    MixedMap map[string]string
}

type BinaryCarrier struct {
    // Inject() should include / Join() should extract Baggage items in
    // Buffer iff BaggageEnabled==true.
    BaggageEnabled bool

    // A buffer in which to store Tracer State and – depending on
    // BaggageEnabled – Baggage items.
    Buffer []byte
}
```

Thoughts? @yurishkuro I'm particularly curious what you think about the general proposal for TChannel above, etc.

**Comment by tschottdorf** [comment]
Passing options in through the carrier types seems pretty fishy to me. Can't you just have two carrier types, one of which doesn't support attributes?

**Comment by yurishkuro** [comment]
@bensigelman 
> I'm particularly curious what you think about the general proposal for TChannel above, etc.
I think I can live with that.

> Can't you just have two carrier types, one of which doesn't support attributes?
Another alternative is to make them completely independent, i.e. instead of the carrier having two parts as before, have two calls, inject tracer state, and inject baggage. An RPC that doesn't want to propagate baggage simply doesn't call the 2nd method (incidentally this solves the TChannel issue via standard carriers again).

**Comment by bg451** [comment]
> instead of the carrier having two parts as before, have two calls, inject tracer state, and inject baggage.

I like this idea. Carriers should carry data and that's it, keep any logic-y stuff out of it.

The current `Inject` function assumes both tracer state and baggage gets injected. It can just calls the two individual functions for injecting state and function. This way you don't break the existing API and get something like this.
```go
func (t *Tracer) Inject(sp Span, format interface{}, carrier interface{}) error
  t.InjectTracerState(sp, format, carrier)
  t.InjectBaggage(sp, format, carrier)
}

func (t *Tracer) InjectTracerState(sp Span, format interface{}, carrier interface{}) error {
  // Inject tracer state the Span ID and sampled status
}

func (t *Tracer) InjectBaggage(sp span, format interface{}, carrier interface{}) error {
  // Inject the baggage
}
```

**Comment by bensigelman** [comment]
@tschottdorf fair point about keeping logic out of the carriers...

FWIW, there is also something similar to what @michaelsembwever proposed with this implementation helper PR in Java: https://github.com/opentracing/opentracing-java/pull/11/files#diff-f2ad9c00db83aa2576fe12ea2f7f0413R21 ... the part I'm really pointing to is that the Tracer itself could be configured to always support XOR ignore the baggage.

@yurishkuro's idea to create separate methods for injecting tracer state and baggage makes some sense to me... then again, `Join(...)` is trickier. @yurishkuro, I think `Join(...)` would run into trouble if different formats were used for the tracer state and baggage; this makes me think we want to strongly encourage the caller to do the injection in a single method call with a single carrier instance. Given that idea, here's yet another way to make this work:

```
type Tracer interface {
...
    Inject(sp Span, format interface{}, carrier interface{}) error
    InjectWithOptions(sp Span, format interface{}, carrier interface{}, InjectOptions opts) error
...
}

type InjectOptions struct {
    IncludeTracerState bool
    IncludeBaggage bool
}
```

**Comment by bg451** [comment]
> different formats were used for the tracer state and baggage

I can't really imagine a case where this would happen, would you mind elaborating?

**Comment by bensigelman** [comment]
@bg451 

> > different formats were used for the tracer state and baggage
>
> I can't really imagine a case where this would happen, would you mind elaborating?

Sure... @yurishkuro was saying that, in `TChannel`, the tracer state would be embedded in binary form at some fixed offset in the RPC body, but the baggage would be key:value string pairs.

**Comment by bg451** [comment]
Ah okay. I didn't catch that in previous discussions, thanks!

**Comment by bensigelman** [comment]
Not sure where we are with this... I was thinking more about the above and realized that Baggage can still be opt-in without touching inject/join. An individual Tracer implementation can simply not-support Baggage, or it would even be possible to write a pass-through Tracer adapter that basically just delegates everything but drops all calls to `SetBaggageItem(...)`.

Having done something like that, I think we'd be better off with the following two required carriers:

```
// SharedMapCarrier is a generic key:value text map in which to store/smuggle
// Tracer State and Baggage.
//
// IMPORTANT: The contents of this map are *not* restricted to Tracer State and
// Baggage: it is potentially shared with other entirely unrelated things
// (e.g., arbitrary HTTP headers). As such, any carrier map entries must be
// distinguishable in some way (e.g., via a shared prefix string).
type SharedMapCarrier map[string]string

// BinaryCarrier is an opaque binary container in which to store Tracer State and
// Baggage.
type BinaryCarrier bytes.Buffer
```

`Inject()` and `Join()` would stay exactly as they are today. Thoughts?

**Comment by yurishkuro** [comment]
> An individual Tracer implementation can simply not-support Baggage, or it would even be possible to write a pass-through Tracer adapter that basically just delegates everything but drops all calls to SetBaggageItem(...)

I am not sure I buy this scenario. It's not about what the tracer supports, but what a composition of different frameworks in a service requires. Imagine a service that talks to some 3rd party storage systems. Not only they may not have any use for you app's baggage, but they also may have highly optimized proprietary protocol that doesn't even support baggage, only some restricted form of tracing context. However, the same service may talk to other services in the stack that do support baggage. So the same tracer will be used by all these RPC frameworks, but with different requirements and different carriers.

**Comment by bensigelman** [comment]
If services in the middle of the stack choose not to support Baggage, then Baggage doesn't really *work*... the whole point is that it cuts across the stack, right?

BTW, yet another possibility would be a pair of baggage-less formats... the IPC programmer would control whether the carrier includes baggage by requesting the `SharedMapCarrier` XOR the `SharedMapNoBaggageCarrier` (would need a better name).

**Comment by yurishkuro** [comment]
What I meant was service A calling B, and service A calling storage S. Both A and B support baggage, but the storage S does not (protocol limitations, or lack of plugin framework in S that make baggage unusable there). So the same tracer in A must deal with injecting a span with and without storage.

It feels like of all the alternatives the simplest would be to expose a parameter in the Inject method:
```go
func Inject(span Span, carrier interface{} includeBaggage bool)
```

I don't believe this has to be symmetrical with the extractor, because extractor just takes what has been transmitted over IPC - (A) if the baggage was transmitted, it will read it from carrier and store in the span, (B) if the baggage was not transmitted, it will keep the baggage empty. I am not sure what benefit we have by _not_ extracting the baggage in (A).

**Comment by bensigelman** [comment]
@yurishkuro IMO the less we can add special parameters/flags for baggage, the better. Going back to your example, if `S` doesn't support baggage, can't it just not-use (or not-understand) the baggage data sent in by `A`? Or, in the unfortunate case where `S` requires its own custom carrier – OT compatibility bedamned – that carrier could exclude baggage entirely.

I just have the sense that we are talking about a really obscure corner case, and furthermore that there are plenty of plausible ways to make that corner case working without introducing a baggage-specific param/flag at `Inject`-time. And in the process we could make the carriers structurally simpler, perhaps reducing to nothing more than a string map and a byte buffer.

**Comment by yurishkuro** [comment]
Fair enough, I concede that this may be an obscure corner case that can be dealt with via custom carrier. Btw, out spec does not show examples of custom using custom carriers, leaving that API completely up to the tracer implementation. The downside of that is swapping the tracers becomes not an `O(1)`.

**Comment by michaelsembwever** [comment]
> I was thinking more about the above and realized that Baggage can still be opt-in without touching inject/join. An individual Tracer implementation can simply not-support Baggage, or it would even be possible to write a pass-through Tracer adapter that basically just delegates everything but drops all calls to SetBaggageItem(...).

I am behind this sentiment.

> furthermore that there are plenty of plausible ways to make that corner case working without introducing a baggage-specific param/flag at Inject-time. 

Yes, and that's what I've demonstrated in https://github.com/opentracing/opentracing-java/pull/11/files#diff-f2ad9c00db83aa2576fe12ea2f7f0413R21

There's plenty of room to solve these problems behind-the-scenes. Additional flags or methods are not required (in the API).

We still have a problem with the serialisation of required carriers needing a specification so that different implementations in different languages, given the a fixed rpc/http framework, deserialise in a compatible way regardless of who serialised.

**Comment by yurishkuro** [comment]
> We still have a problem with the serialisation of required carriers needing a specification so that different implementations in different languages, given the a fixed rpc/http framework, deserialise in a compatible way regardless of who serialised.

I don't see why we need such specification. There is no goal of making two tracing systems (e.g. Zipkin vs. Dapper) inter-operate via serialization spec. The only requirement is the same tracing implementation (even in different languages, e.g. Zipkin Java vs. Zipkin Python) must understand its own serialization format.

**Comment by michaelsembwever** [comment]
> The only requirement is the same tracing implementation (even in different languages, e.g. Zipkin Java vs. Zipkin Python) must understand its own serialization format. 

That is correct.

> Which becomes the problem for that implementation, not for OpenTracing spec.

But what is the common space here?
Tracing implementations are just going to push the problem back onto OpenTracing. This is about how *OpenTracing serialisation* works for each of the different Tracing solutions. With OpenTracing serialisation you could well already have broken the vanilla Tracing serialisation (eg the need to separate state and baggage). Therefore such specification does not belong to the Tracing solution, but to OpenTracing instead. If the code lives in OpenTracing this is even more evident. If OpenTracing is also delegating the serialisation into the rpc/http framework back onto the Tracing own instrumentation libraries, then you could say the common space is then within that Tracing solution. In this situation it is easier for the OpenTracing Internal Specfication to simply reference the Tracing solution's documentation.

At the end of the day I suspect it'll be more pragmatic to let OpenTracing provide a cohesive set of docs, re-using (referencing) what it can, rather than trying to force others to do what we think is correct.

**Comment by yurishkuro** [comment]
We can certainly write an "OpenTracing opinion" about how the serialization could be done, but we cannot, nor should we IMO, enforce how it's done. I disagree that OT breaks the existing serialization, precisely because it does not dictate how it's done. OT tracer that wants to be compatible with Finagle can continue encoding trace context as 4 B3_xxx headers.

I may be misunderstanding what you are trying to suggest. The existing k/v text carrier is meant to be tracer-specific encoding. The RPC framework is expected to store those k/v pairs in the "application headers" segment of the request (e.g. HTTP headers) without any additional encoding or serialization.

**Comment by bensigelman** [comment]
In the spirit of keeping the discussion on-topic, I wonder if we could fork off a separate issue about whether OpenTracing should (or must!) have an opinion about how implementations do serialization. If we agree that such a discussion can happen independent of the actual issue here – namely, what the required carriers should look like – I will create a new issue and these two threads can happen in parallel.

In the meantime, I will go back to @yurishkuro's (on-topic) message above:

> Fair enough, I concede that this may be an obscure corner case that can be dealt with via custom carrier.

So you are ok with the (simpler) single text map and single byte buffer carriers (at least for v1.0RC2)?

> Btw, out spec does not show examples of custom using custom carriers, leaving that API completely up to the tracer implementation. The downside of that is swapping the tracers becomes not an `O(1)`.

There are some typos above which might be getting in the way of my understanding, but I think you're pointing out that our docs/spec doesn't give any guidance about custom carriers... My hope was that code that relies on custom carriers would fall back on one of the required carriers if the tracing impl didn't support the non-standard/custom carrier. Imagine a pirate-friendly RPC framework called "ArrrrPC" – pseudocode might look like this:

```
func sendArrrrPC(rpcHandle, span) {
  try {
    // First try the custom ArrrrPC format
    span.tracer.inject(span, arrrrpc.ARRRRPC_FORMAT, rpcHandle);
  } catch (opentracing.UnsupportedFormat) {
    // Fall back on TEXT_FORMAT as needed
    textCarrier = opentracing.TextCarrier();
    span.tracer.inject(span, opentracing.TEXT_FORMAT, textCarrier);
    rpcHandle.headers.merge(textCarrier);
  }
  ... actually send the RPC ...
}
```

If the above makes sense, I agree it should be part of the docs.

**Comment by yurishkuro** [comment]
Apologies for the typos, let me reword:
> our spec does not show examples of _registering_ custom carriers, leaving that API completely up to the tracer implementation. The downside of that is swapping the tracers becomes not an `O(1)`. 

Re the ARRRRPC_FORMAT example (love the name), the assumption is that its requirements can be satisfied, perhaps less efficiently, by standard formats. But as we have seen in TChannel use case, no standard carrier can be used as a fallback there, _especially_ if we decide not to separate state and baggage. 

Anyway, I don't think this is, strictly speaking, on-topic either.

**Comment by michaelsembwever** [comment]
> if we could fork off a separate issue about whether OpenTracing should (or must!) have an opinion about how implementations do serialization. 

Yes. FTR I never advocated for OpenTracing to dictate how implementations must serialise, but allow implementations to update OpenTracing docs along with OpenTracing code, as they implement.

**Comment by michaelsembwever** [comment]
> But as we have seen in TChannel use case, no standard carrier can be used as a fallback there, especially if we decide not to separate state and baggage. 

In the delegation model proposed in opentracing-java it was intentional that standard carriers did not have to be used. (That's kind of the idea with a delegation model).

**Comment by yurishkuro** [comment]
@michaelsembwever even with delegation, the assumption is that a custom handler is delegating to some "standard" handlers via some "standard formats". My point is that if the standard formats are (1) a single map, and (2) a single byte array, then neither of these can be made to work for TChannel, which requires a fixed-length byte array for trace context, and a string k/v map for baggage.

**Comment by michaelsembwever** [comment]
> the assumption is that a custom handler is delegating to some "standard" handlers via some "standard formats".

since when? 
this sounds like the spec trying to stick its nose into the implementation.

**Comment by yurishkuro** [comment]
I think we're disconnecting again. Who is implementing the delegation in your scenario, the end user or the tracer implementor?  I am talking about the end user (e.g. the author of TChannel RPC). Since the end user has no knowledge about tracer internal, the best she can do when implementing a custom injector is to delegate to known/standard formats that are guaranteed to be supported by the tracer.

**Comment by michaelsembwever** [comment]
> Since the end user has no knowledge about tracer internal… 

If the end-user working with a custom rpc (that has no known/standard format) is expected to solve this without knowing of the tracer internals I think we're kinda screwed.

The normal situation is that end-users delegate to a known/standard format.
There arises situations where the end-user can't, and must accordingly implement the injector/extractor without such delegation. TChannel seems a good example of this.

**Comment by yurishkuro** [comment]
> The normal situation is that end-users delegate to a known/standard format. There arises situations where the end-user can't, and must accordingly implement the injector/extractor without such delegation.

That's exactly what I was saying. However, in the existing Go API where text/binary carriers are "split", it _was_ possible to implement TChannel use case via delegation (just not very efficiently).

@bensigelman as I said earlier, I don't have strong objection to "un-splitting" the carriers.

**Comment by bensigelman** [comment]
Alright... I will wait to see if there are immediate objections to the great un-splitting. If not we can start a Github Issue timer and go from there.

**Comment by bensigelman** [comment]
If there are no objections within 48 hours, we will proceed with the two required carriers described above ([`SharedMapCarrier` and `BinaryCarrier`](https://github.com/opentracing/opentracing.github.io/issues/73#issuecomment-192752866)).

**Comment by bensigelman** [comment]
Ok, no objections... I will start working through this over the weekend (hopefully).
